source/acquaman/BioXASSide/BioXASSideXRFScanController.cpp

```diff
@@ -0,0 +1,143 @@
+
+		scan_->rawData()->addMeasurement(AMMeasurementInfo(QString("raw%1").arg(i+1), QString("Element %1").arg(i+1), "eV", detector_->axes()));
+		scan_->addRawDataSource(new AMRawDataSource(scan_->rawData(), i), true, false);
+	}
+
+	AM1DSummingAB *summedSpectra = new AM1DSummingAB("SummedSpectra");
+	QList<AMDataSource *> list;
+
+	for (int i = 0; i < scan_->rawDataSourceCount(); i++)
+		list << (AMDataSource *)scan_->rawDataSources()->at(i);
+
+	summedSpectra->setInputDataSources(list);
+	scan_->addAnalyzedDataSource(summedSpectra, true, false);
+}
+
+BioXASSideXRFScanController::~BioXASSideXRFScanController()
+{
+
+}
+
+void BioXASSideXRFScanController::onProgressUpdate()
+{
+	emit progress(detector_->elapsedTime(), detector_->acquisitionTime());
+}
+
+void BioXASSideXRFScanController::onStatusChanged()
+{
+	if (!detector_->isAcquiring())
+		onDetectorAcquisitionFinished();
+}
+
+bool BioXASSideXRFScanController::startImplementation()
+{
+	connect(detector_, SIGNAL(acquisitionStateChanged(AMDetector::AcqusitionState)), this, SLOT(onStatusChanged()));
+	connect(detector_, SIGNAL(elapsedTimeChanged(double)), this, SLOT(onProgressUpdate()));
+	detector_->acquire();
+
+	setStarted();
+
+	return true;
+}
+
+void BioXASSideXRFScanController::onDetectorAcquisitionFinished()
+{
+	disconnect(detector_, SIGNAL(acquisitionStateChanged(AMDetector::AcqusitionState)), this, SLOT(onStatusChanged()));
+	disconnect(detector_, SIGNAL(elapsedTimeChanged(double)), this, SLOT(onProgressUpdate()));
+
+	QVector<double> spectrum = QVector<double>(detector_->size(0));
+
+	for (int i = 0, elements = detector_->elements(); i < elements; i++){
+
+		detector_->rawSpectrumSources().at(i)->values(AMnDIndex(0), AMnDIndex(detector_->size(0)-1), spectrum.data());
+		scan_->rawData()->setValue(AMnDIndex(), i, spectrum.constData());
+	}
+
+	// MB: Notice for Darren: at setFinished(), AMScanController will automatically ensure scan()->setScanController(0), and this will trigger a change in the scan's db state. (currentlyScanning()).
+	// If you want this to not create a prompt in AMGenericScanEditor ("The scan has been modified... Do you want to save it?") can do this first:
+	scan()->setScanController(0);
+	// before saving to the DB.
+
+	saveData();
+
+	if(scan()->database())
+		scan()->storeToDb(scan()->database());
+
+	else
+		scan()->storeToDb(AMDatabase::database("user"));
+
+	setFinished();
+}
+
+void BioXASSideXRFScanController::saveData()
+{
+	QFile file(AMUserSettings::userDataFolder + "/" + scan_->filePath());
+	if(!file.open(QIODevice::WriteOnly)) {
+		AMErrorMon::error(this, -1, "Could not save XRF data.");
+		return;
+	}
+
+	QTextStream out(&file);
+
+	// There are two types of raw data sources.  The spectra of rank 1 and the dead time of rank 0.
+	for (int i = 0; i < scan_->rawDataSources()->count(); i++){
+
+		if (scan_->rawDataSources()->at(i)->rank() == 1){
+
+			qint32 counts = scan_->rawData()->value(AMnDIndex(), i, AMnDIndex(0));
```

Comment:
It probably should be quint, but I've been using qint32 for more than three years and it's been fine so far.  Should be fine for the time being.